pkg/intermediate/aggregate.go

```diff
@@ -124,74 +140,272 @@ func (a *AggregationProcess) DeleteFlowKeyFromMap(flowKey FlowKey) {
 			}
 		}
 	}
-	a.addRecordToMap(flowKey, record)
-	a.removeDuplicates(flowKey)
 }
 
-// removeDuplicates is currently used only in correlateRecords().
-// For other uses, please acquire the flowKeyRecordLock for protection.
-func (a *AggregationProcess) removeDuplicates(flowKey FlowKey) {
-	records := a.flowKeyRecordMap[flowKey]
-	srcRecords := make([]entities.Record, 0)
-	dstRecords := make([]entities.Record, 0)
-	for _, record := range records {
-		if isRecordFromSrc(record) {
-			srcRecords = append(srcRecords, record)
-		} else {
-			dstRecords = append(dstRecords, record)
+// aggregateRecords aggregate the incomingRecord with existingRecord by updating
+// stats and flow timestamps.
+func (a *AggregationProcess) aggregateRecords(incomingRecord, existingRecord entities.Record, fillSrcStats, fillDstStats bool) {
+	if a.aggregateElements == nil {
+		return
+	}
+
+	for _, field := range a.aggregateElements.nonStatsElements {
+		if ieWithValue, exist := incomingRecord.GetInfoElementWithValue(field); exist {
+			switch ieWithValue.Element.Name {
+			case "flowEndSeconds":
+				existingIeWithValue, _ := existingRecord.GetInfoElementWithValue(field)
+				// Update flow end timestamp if it is latest.
+				if ieWithValue.Value.(uint32) > existingIeWithValue.Value.(uint32) {
+					existingIeWithValue.Value = ieWithValue.Value
+				}
+			default:
+				klog.Errorf("Fields with name %v is not supported in aggregation fields list.", field)
+			}
 		}
 	}
-	if len(srcRecords) != 0 {
-		a.flowKeyRecordMap[flowKey] = srcRecords
-	} else {
-		a.flowKeyRecordMap[flowKey] = dstRecords
+
+	statsElementList := a.aggregateElements.statsElements
+	antreaSourceStatsElements := a.aggregateElements.aggregatedSourceStatsElements
+	antreaDestinationStatsElements := a.aggregateElements.aggregatedDestinationStatsElements
+	for i, element := range statsElementList {
+		isDelta := false
+		if strings.Contains(element, "Delta") {
+			isDelta = true
+		}
+		if ieWithValue, exist := incomingRecord.GetInfoElementWithValue(element); exist {
+			existingIeWithValue, _ := existingRecord.GetInfoElementWithValue(element)
+			// Update the corresponding element in existing record.
+			if !isDelta {
+				existingIeWithValue.Value = ieWithValue.Value
+			} else {
+				// We are simply adding the delta stats now. We expect delta stats to be
+				// reset after sending the record from flowKeyMap in aggregation process.
+				// Delta stats from source and destination nodes are added, so we will have
+				// two times the stats approximately.
+				// For delta stats, it is better to use source and destination specific
+				// stats.
+				existingIeWithValue.Value = existingIeWithValue.Value.(uint64) + ieWithValue.Value.(uint64)
+			}
+			// Update the corresponding source element in antreaStatsElement list.
+			if fillSrcStats {
+				existingIeWithValue, _ := existingRecord.GetInfoElementWithValue(antreaSourceStatsElements[i])
```

Comment:
Yes, I would suggest to make statsElementList, antreaSourceStatsElements, and antreaDestinationStatsElements all independent string lists. Also it looks like there should be no duplicated elements in these three lists, otherwise the stats could be double or triple counted?

Reply:
Yes this part need to have more checks that validate the assumptions taken in the code. As Antrea is the only user and we control the input. 
I want to add a TODO and move on, considering the time factor. What do you say?

Reply:
I'm ok with adding TODO comment.

Reply:
OK too